### PR TITLE
Next: Add Ratings Component

### LINKED ISSUE #2392: NEXT: Ratings
This will act as a hub to centralize this information.

## Maintainer Requests

The following requests are coming straight from the Skeleton team. These are highly likely be implemented:

- [x] Should adhere to the [ARIA APG](https://www.w3.org/WAI/ARIA/apg/patterns/radio/examples/radio-rating/)
- [x] In the v2 version we had a "wobble" due to the position of the Font Awesome star icons. Lucide should not suffer from this, but let's pay close attention to ensure this is not the case in v3.
- [x] Recommend a smaller display size by default in the docs. The icons were way too big in v2!
- [x] (optional) [Fractional support](https://mantine.dev/core/rating/#fractions) is interesting, but not required

## Community Requests

The following requests have come from the community and are under consideration:
 
* https://github.com/skeletonlabs/skeleton/issues/2558

## Bugs and Issues

* n/a

## References

- Skeleton v2: https://www.skeleton.dev/components/ratings
- Mantine: https://mantine.dev/core/rating/

## Feedback

If you have additional updates or requests for this feature, please do so in the comments section below.

### PR DESCRIPTION
**Pull request by Mahmoud-zino**
## Linked Issue

Closes #2392

### CODE REVIEW

**Comment by endigo9740** [comment]
@Mahmoud-zino just took a quick look at this. I won't provide a full review as I know it's still a draft - but make sure you're not hardcoding SVGs within components if it can be avoid. While we're taking an opinionated stance on Lucide, that doesn't mean we're keen to embed icons from it.

**Comment by Mahmoud-zino** [comment]
> @Mahmoud-zino just took a quick look at this. I won't provide a full review as I know it's still a draft - but make sure you're not hardcoding SVGs within components if it can be avoid. While we're taking an opinionated stance on Lucide, that doesn't mean we're keen to embed icons from it.

This is just a fallback svg (star since it is the most common one for rating) but if you want to use another approach I am totally ok with it, just let me know what would you prefer.

**Comment by endigo9740** [comment]
Yeah, no fallback SVG. Users will be required to add their own icons. Placeholder text is probably fine though.

**Comment by Mahmoud-zino** [comment]
@endigo9740 Please ignore the fallback SVG at the moment, I want to get your opinion on the fractional approach before continuing the implementation.
so this approach is using clip-path to hide parts of the full icon to display the empty icon under it, this will enable us to support fractions fully with ease, you can look at the example svelte file to see it in action...
please let me know if I should continue this approach or you have something else in mind

**Comment by endigo9740** [comment]
@Mahmoud-zino `clip-path` is a great way to handle this! I like it. Let's do it!

**Comment by endigo9740** [comment]
@Mahmoud-zino as part of prepping for tomorrow's release I'm doing a pass through all pending PRs. Let me know a status update on this one. What's remaining and ETA for delivery. If you're bogged down, let me or the team know and someone can likely jump in and pick this up! (Hugo, etc)

**Comment by Mahmoud-zino** [comment]
@endigo9740 Functionality and design of the component is now implemented and ready for review :+1:

**Comment by endigo9740** [comment]
@Mahmoud-zino I'm so sorry, I've been swamped with the Floating UI Svelte stuff this week and completely missed this.

I'll aim to review in the next day or two, could you go ahead and do me a huge favor - could you install Lucide for the Svelte and React packages as dev dependencies. Then go ahead and swap out SVGs with icons from those?

Remember, we will NEVER used Lucide in our components, but it's perfectly valid to use in the SvelteKit and Vite/React sandboxes. I've been meaning to do this for a while, and it'll be helpful for your component here.

Please and thanks sir!

**Comment by endigo9740** [comment]
Thanks @Mahmoud-zino this is on my agenda to review today, assuming no further interruptions!

File: packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte
Comment (endigo9740) [inline packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte:104]: Let's make the wrapping element a `<figure>` instead of div.

File: packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte
Comment (endigo9740) [inline packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte:105]: Let's drop the `role` here. The rule of thumb with a11y is no-a11y attributes are better than bad or incorrect a11y attributes. This will become a Figure role per the above comment.

File: packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte
Comment (endigo9740) [inline packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte:112]: I'm not sure I understand the extra wrapping `div` here. Overall structure should be:

```html
<figure>
    <button /> (or div)
    <!-- repeat per instance -->
</figure>
```

File: packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte
Comment (endigo9740) [inline packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte:108]: Not sure I understand the extra events on the parent. Neither `<div>` or `<figure>` are directly interactive, which is why I'm assuming a role was given. But this is typically an indicator things are not setup properly. Keep interaction limited to interactive elements only.

For example, when I hover or click a button, handle the interaction there only.

File: packages/skeleton-svelte/src/lib/components/Rating/types.ts
Comment (endigo9740) [inline packages/skeleton-svelte/src/lib/components/Rating/types.ts:37]: It's easier to identify related snippets with prefixes versus suffixes. At least in LTR languages! :)

- `iconEmpty`
- `iconFull`

File: packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte
Comment (endigo9740) [inline packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte:65]: There's a lot of boilerplate being replicated between interactive vs non-interactive.

I'm trying to think of a way to handle this. Maybe look into `<svelte:element>`, which lets you conditionally set the HTML tag to use:

https://svelte.dev/docs/special-elements#svelte-element

The issue would be dynamically applying the attributes and event handlers. Maybe you could use a spread operator like:

```html
<svelte:element bind:this={elemTag} {...attributes}>...</ svelte:element>
```

File: packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte
Comment (endigo9740) [inline packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte:53]: Looks like some logic is missing from the `else` condition?

File: packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte
Comment (endigo9740) [inline packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte:33]: Console log was committed

File: packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte
Comment (endigo9740) [inline packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte:18]: The `aspect-square` Tailwind utility class is equivalent to 1:1

File: packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte
Comment (endigo9740) [inline packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte:74]: There's a ton of inline classes that should likely be abstracted to `baseX` style props.

File: packages/skeleton-svelte/src/lib/components/Rating/types.ts
Comment (endigo9740) [inline packages/skeleton-svelte/src/lib/components/Rating/types.ts:31]: To just to confirm, icon size is set externally, correct? We don't need an `itemSize` attribute, is that right?

https://tailwindcss.com/docs/size

File: packages/skeleton-svelte/src/routes/components/ratings/+page.svelte
Comment (endigo9740) [inline packages/skeleton-svelte/src/routes/components/ratings/+page.svelte:102]: Typo: "Interactive"

**Review by endigo9740 — CHANGES_REQUESTED** [review]
@Mahmoud-zino here's my review. A few things of note:

- I LOVE the fractional units. That's so much cleaner.
- I'm kind of so-so on the hover states, that may be overkill and adding extra complexity
- Keep the interaction though. Clicking to set state is nice.

Let me know if I can clarify any points raised below. I'd suggest we go through another round of review before proceeding to tests and React components.

File: packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte
Comment (Mahmoud-zino) [reply packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte:53]: We agreed last time to handle keyboard logic after the first review, and I have some questions about how are we going to handle fractions with Keyboard, so I would wait until the first wave of changes are approved :+1:

File: packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte
Comment (Mahmoud-zino) [reply packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte:74]: I don't think it makes sense to abstract those classes, they are crucial for the functionality of the component and should not be changed by the user IMHO.
And I am not sure what to call the element if abstracted, maybe you have a suggestion...

**Comment by endigo9740** [comment]
@Mahmoud-zino I've bumped this back to a draft to help better indicate when it's ready for review again!

**Comment by Mahmoud-zino** [comment]
@endigo9740 I have left you a couple of comments back that needs some clarifications :smile: Thanks.
Also about the hover, I removed it but I am really thinking we should keep it, it is a lot easier to select fractions when you are getting immediate feedback.

> If we can reduce complexity enough, I'd like us to consider inserting input state so this can be used in forms!

Do you mean adding a prop so users can change the type of the buttons to submit a form? if not, could you please provide more details?

**Comment by endigo9740** [comment]
@Mahmoud-zino this is still on my list to review. Just tied up with the auto-doc stuff atm.

> Do you mean adding a prop so users can change the type of the buttons to submit a form? if not, could you please provide more details?

I believe buttons and inputs cannot have their type dynamically set. Svelte will complain about this. In this case I think we should rely on users to handle submit themselves and keep the buttons at `type="button"`

File: packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte
Comment (endigo9740) [reply packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte:45]: The problem her is we can't promote the library as having faux-headless support if some classes are hardcoded. Even if they're structurally important, consider moving them to a `base` prop of some sort.

File: packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte
Comment (Mahmoud-zino) [reply packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte:36]: @endigo9740 Do you have any suggestions on how to handle the keyboard events with fractions? do we ignore the fractions or use something like arrow to set the different fractions?

Comment (endigo9740) [reply packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte]: @Mahmoud-zino I like how Mantine handles this. Provide focus, then use left/right arrows:
https://mantine.dev/core/rating/#fractions

**Comment by endigo9740** [comment]
@Mahmoud-zino hopefully you don't murder me for taking the initiative here, but I've completed a PR review and a bit of a refactor to the component. This streamlines a lot of the complexity and uses much less code for the same results.

Notable changes include:

- The local embedded snippet is gone; code has been moved within the each block
- The `<svelte:element>` in gone in favor of a hardcoded button
- The button hover state is set by `pointer-events-none` when `interactive` is false.
- I've refactored the styles to confirm to the base/dynamic/classes pattern ([see dynamic reference](https://next.skeleton.dev/docs/resources/contribute/components#dynamic-style-props)). There should be less redundancy.
- I've change the semantics for a few of the prop names
- Fixed an issue where snippets names internally/externally did not match

My hope here is that by slimming down the complexity, it should make it easier to focus on adding keyboard interactions. That would be the final piece before this ready to create documentation and port to React.

Let me know what you think!

**Comment by Mahmoud-zino** [comment]
@endigo9740 I just reviewed your changes and I am really happy with what you did :+1: 
also, I just implemented the keyboard interaction part, and here are some points to notice about that:
1. I had to change the `onclick` event to `onmousedown` event so pressing enter won't miss with the calculation done on the mouse position when clicking for fractions.
2. for some reason I couldn't get the tabIndex to update re-actively so I've gone the ref route here.

Let me know what you think about the changes and if you wish to go another route for the keyboard interaction.

**Comment by endigo9740** [comment]
@Mahmoud-zino sounds good. I'd say go ahead and setup a doc page for this and that'll make it easier for me to preview over the weekend (if you've got time!). Otherwise I'll review on Monday.

**Comment by endigo9740** [comment]
@Mahmoud-zino thank you for adding the doc page. Things are looking great! With a quick glance, I'm good with how you're handling the keyboard events. The only suggestion I might have would be to default to `ltr` instead of `rtl`. The majority of the project, maintainers, and likely contributors will be able to make more sense of that.

**Comment by endigo9740** [comment]
@Mahmoud-zino I saw this was still marked "ready for review" so I've bumped this back to the "draft" state after a quick review. I modified a couple of the doc pages, but otherwise everything in the component looks fine. You're good to roll with the React version asap!

**Comment by endigo9740** [comment]
@Mahmoud-zino just saw this marked as ready for review. I'll aim for a PR review first thing tomorrow. Thanks!

File: packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte
Comment (endigo9740) [inline packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte:12]: The `-token` classes no longer exist in v3. We should also avoid hardcoding values whenever possible.

If you were going automatically set these I'd opt for something like:

> Current means inherit from the nearest parent, aka "use the default"

```ts
text = 'text-current',
fill = 'fill-current',
```

Optionally, when providing set values, always use a pairing like: `fill-surface-950-50`.

However, users are supplying their own icons, so perhaps they should be responsible for these values? If so, drop these two props. Honestly this would be my recommendation. Less is more.

File: sites/next.skeleton.dev/src/examples/components/ratings/Example.svelte
Comment (endigo9740) [inline sites/next.skeleton.dev/src/examples/components/ratings/Example.svelte:8]: Remove the hardcoded white fill. Either use the default (per my other comment) or use a color pairing like `class="fill-surface-950-50"`. Currently this is impossible to see in light mode:

Screenshot 2024-06-11 at 2 06 13 PM

Screenshot 2024-06-11 at 2 06 19 PM

File: sites/next.skeleton.dev/src/examples/components/ratings/Example.tsx
Comment (endigo9740) [inline sites/next.skeleton.dev/src/examples/components/ratings/Example.tsx:12]: Another hardcoded fill, switch to: `class="fill-surface-950-50"`

File: sites/next.skeleton.dev/src/examples/components/ratings/FractionExample.svelte
Comment (endigo9740) [inline sites/next.skeleton.dev/src/examples/components/ratings/FractionExample.svelte:8]: Another hardcoded fill, switch to: `class="fill-surface-950-50"`

File: sites/next.skeleton.dev/src/examples/components/ratings/FractionExample.tsx
Comment (endigo9740) [inline sites/next.skeleton.dev/src/examples/components/ratings/FractionExample.tsx:13]: Another hardcoded fill, switch to: `class="fill-surface-950-50"`

File: sites/next.skeleton.dev/src/examples/components/ratings/RTLExample.svelte
Comment (endigo9740) [inline sites/next.skeleton.dev/src/examples/components/ratings/RTLExample.svelte:11]: Another hardcoded fill, switch to: `class="fill-surface-950-50"`

File: sites/next.skeleton.dev/src/examples/components/ratings/RTLExample.tsx
Comment (endigo9740) [inline sites/next.skeleton.dev/src/examples/components/ratings/RTLExample.tsx:14]: Another hardcoded fill, switch to: `class="fill-surface-950-50"`

File: sites/next.skeleton.dev/src/examples/components/ratings/UserInteractiveExample.svelte
Comment (endigo9740) [inline sites/next.skeleton.dev/src/examples/components/ratings/UserInteractiveExample.svelte:8]: Another hardcoded fill, switch to: `class="fill-surface-950-50"`

File: sites/next.skeleton.dev/src/examples/components/ratings/UserInteractiveExample.tsx
Comment (endigo9740) [inline sites/next.skeleton.dev/src/examples/components/ratings/UserInteractiveExample.tsx:13]: Another hardcoded fill, switch to: `class="fill-surface-950-50"`

File: packages/skeleton-react/src/lib/components/Rating/Rating.tsx
Comment (endigo9740) [inline packages/skeleton-react/src/lib/components/Rating/Rating.tsx:41]: Keep your `useState` definitions grouped at the top, similar to `$state()` runes or vars in general.

File: packages/skeleton-react/src/lib/components/Rating/Rating.tsx
Comment (endigo9740) [inline packages/skeleton-react/src/lib/components/Rating/Rating.tsx:66]: Remove the whitespace here (the line return). It'll help readability of the code.

File: packages/skeleton-react/src/lib/components/Rating/Rating.tsx
Comment (endigo9740) [inline packages/skeleton-react/src/lib/components/Rating/Rating.tsx:100]: Unless React has a specific implementation, I'd recommend keeping `<style>` blocks at the bottom of the template to match Svelte, Vue, etc

File: packages/skeleton-react/src/lib/components/Rating/Rating.tsx
Comment (endigo9740) [inline packages/skeleton-react/src/lib/components/Rating/Rating.tsx:127]: I think the dedicated `IconEmpty` and `IconFull` components may be overkill. For simple "slot-like" content (such as icons) React allows you to pass a `ReactNode` (read: React components themselves) as if they are props...

Types:
https://github.com/skeletonlabs/skeleton/blob/next/packages/skeleton-react/src/lib/components/Switch/types.ts#L63

Props:
https://github.com/skeletonlabs/skeleton/blob/next/packages/skeleton-react/src/lib/components/Switch/Switch.tsx#L42

Template Usage:
https://github.com/skeletonlabs/skeleton/blob/next/packages/skeleton-react/src/lib/components/Switch/Switch.tsx#L96

Example:
https://github.com/skeletonlabs/skeleton/blob/next/sites/next.skeleton.dev/src/examples/components/switch/ExampleIcons.tsx#L15

The end result will be a singular component that's much similar to the Svelte implementation! I realize this will take a non-trivial refactor, so let me know if you need help with this!

**Review by endigo9740 — CHANGES_REQUESTED** [review]
Here's my review comments @Mahmoud-zino. This is a mix of small, medium, and large changes. Please review each and let me know if I can clarify anything.

Given the scope of some of these, I may be able to make myself available the next few days to help if you need it. Just let me know!

File: packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte
Comment (Mahmoud-zino) [reply packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte:112]: yeah unfortunately this doesn't work on DOM elements :cry:

File: packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte
Comment (Mahmoud-zino) [reply packages/skeleton-svelte/src/lib/components/Rating/Rating.svelte:90]: Option 1 is so nice! I had to make some adjustments but it is working as expected, thank you for the cool tip :+1:

File: packages/skeleton-react/src/lib/components/Rating/Rating.tsx
Comment (Mahmoud-zino) [reply packages/skeleton-react/src/lib/components/Rating/Rating.tsx]: Per the solution suggested above, there is no `style` block anymore :partying_face:

**Comment by Mahmoud-zino** [comment]
@endigo9740 I applied all of your suggestions, thank you for your patience and always coming with cool suggestions :+1:

**Comment by endigo9740** [comment]
@Mahmoud-zino and thanks for your help! I'm going to *try* to find some time to squeeze in a follow up review between now and our release on Tuesday, but no promises. Ideally I'd love to launch our next v3 pre-release with this, my upcoming nav components, and the recent progress ring component.

So feel free to polish anything anything you can in the meantime, and I'll ping you before I begin my review (likely Monday).

**Comment by endigo9740** [comment]
@Mahmoud-zino I noticed the most recent deployment also failed. Make sure to do the following:

1. Delete the local `node_modules` in each project you made changes
2. Also delete the `package-lock.json` in the root of the monorepo
2. Upgrade your PNPM version to `9.3.0` (specifically)
3. Finally, run `pnpm i` to reinstall the local

This will do two things:

1. Ensure you're running the current version of PNPM, which @AdrianGonz97 recently made mandatory
2. Regenerate the lock file to ensure it's reflecting in all changes.

If you have any questions about this, ping me or Adrian this weekend and we can assist.

**Comment by endigo9740** [comment]
@Mahmoud-zino I went ahead and implemented some last minute changes to this component. The only one worth mentioning is a minor refactor the keyboard navigation to support up/down in addition to left/right, per:

https://www.w3.org/WAI/ARIA/apg/patterns/radio/examples/radio-rating/#kbd_label

There's a few more things I'd like to revisit in the future, such as dynamically swapping between button/div based on whether there's interaction or not. But this is not worth holding the component back for now. We'll go ahead and merge and revisit in the future.

Great job on this one!

**Pull request merged**
